### PR TITLE
[core] catch exception in async_callback

### LINKED ISSUE #46452: CI: `test_on_completed_callback_refcount` failed on Python 3.12
### What happened + What you expected to happen

`test_on_completed_callback_refcount` failed on Python 3.12

### Versions / Dependencies

Python 3.12

### Reproduction script

```
python -m pytest -vvs python/ray/tests/test_asyncio.py -k "test_on_completed_callback_refcount"
```

### Issue Severity

None

### PR DESCRIPTION
**Pull request by hongchaodeng**
## Why are these changes needed?

async_callback doesn't catch the exception currently. cython will ignore the exception if any thrown.
This will break the test `test_on_completed_callback_refcount` on Python 3.12 because the traceback refcount is not decremented correctly. Since it is ignoring the exception before, simply logging it will fix the issue and still provide better user experience.

## Related issue number

fix https://github.com/ray-project/ray/issues/46452

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rynewang** [comment]
why is the ref count not decremented since it's in `finally` block?

**Comment by hongchaodeng** [comment]
@rynewang 
That's some Python internal optimization. See the repro script in https://github.com/ray-project/ray/issues/46452
It's a regression due to upgrade.
But catching it is what we should do anyway.

**Comment by hongchaodeng** [comment]
> why is the ref count not decremented since it's in finally block?
@rynewang The finally block is called. That's not the real issue. It's the traceback reference that remains.

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -5164,6 +5164,10 @@ cdef void async_callback(shared_ptr[CRayObject] obj,
 
         user_callback = <object>user_callback_ptr
         user_callback(result)
+    except Exception as e:
```
Comment (rynewang) [inline python/ray/_raylet.pyx:5167]: we can remove `as e` here. Also, can we add a note and better an URL to some pages describing this potential bug on Python&Cython?

**Review by rynewang — APPROVED** [review]

**Review by jjyao — APPROVED** [review]

**Pull request merged**